<a href="https://colab.research.google.com/github/BautistaDavid/Proyectos_ClaseML/blob/corte_1/Proyecto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install wooldridge

## **Proyecto 2**

Se va a construir un objeto para poder hallar algunas estadísticas descriptivas de un vector numérico de datos. Después de esto se va aprobar la clase construida usando algunas variables de la base de datos “wage1” del libro de wooldridge.  

In [ ]:
class estadisticos:
  
  def __init__(self,lst):
    self.lst = lst
    return None

  def media(self): 
    return sum(self.lst) / len(self.lst)

  def test(self):
    return self.media()
  
  def desv_est(self):
    return (sum([(i - self.media())**2 for i in self.lst]) / (len(self.lst)-1))**0.5  

  def varianza(self):
    return self.desv_est()**2

  def mediana(self):
    return sorted(self.lst)[len(self.lst)//2] if len(self.lst)%2 != 0 else (sorted(self.lst)[len(self.lst)//2-1]+sorted(self.lst)[len(self.lst)//2]) / 2
  
  def curtosis(self):
    return  sum([(i - self.media())**4 for i in self.lst]) / (len(self.lst)*self.desv_est()**4)

  def simetria(self):
    return  sum([(i -self.media())**3 for i in self.lst]) / (len(self.lst)*self.desv_est()**3)
  
  def coeficiente_variacion(self):
    return self.desv_est() / abs(self.media()) 

Se va a probar la clase para las variables ```wage``` y ```educ``` de la bse de datos ```wage1```. A continuación se despliega información sobre estos datos. 

In [ ]:
import wooldridge as wd
wd.data("wage1",description = True) #Accedemos a información sobre las variables.

name of dataset: wage1
no of variables: 24
no of observations: 526

+----------+---------------------------------+
| variable | label                           |
+----------+---------------------------------+
| wage     | average hourly earnings         |
| educ     | years of education              |
| exper    | years potential experience      |
| tenure   | years with current employer     |
| nonwhite | =1 if nonwhite                  |
| female   | =1 if female                    |
| married  | =1 if married                   |
| numdep   | number of dependents            |
| smsa     | =1 if live in SMSA              |
| northcen | =1 if live in north central U.S |
| south    | =1 if live in southern region   |
| west     | =1 if live in western region    |
| construc | =1 if work in construc. indus.  |
| ndurman  | =1 if in nondur. manuf. indus.  |
| trcommpu | =1 if in trans, commun, pub ut  |
| trade    | =1 if in wholesale or retail    |
| services | =1 if in services indus.  

In [ ]:
import numpy as np
import pandas as pd 
datos = wd.data("wage1")

for i in ["wage","educ"]:
  stats = estadisticos(datos[i])
  mensaje = f"Error"
  funciones = {"Media":[stats.media(),np.mean(datos[i])],
               "Mediana":[stats.mediana(),np.median(datos[i])],
               "Desviación estandar":[stats.desv_est(),np.std(datos[i])],
               "Varianza":[stats.varianza(),np.var(datos[i])],
               "Coeficiente Variación":[stats.coeficiente_variacion(),np.std(datos[i])/np.mean(datos[i])]}
  print(f"Estadisticas {i}")             
  for key,value in funciones.items():
    try:
      assert value[0]==value[1],mensaje
      print(f"{key} = {value[1]} --> Coincide con Numpy")
    except:
      print(f"{key} = {value[1]} --> No Coincide con Numpy, hay una diferencia de {abs(value[1]-value[0])}")
  print("___________________________________________\n ")


Estadisticas wage
Media = 5.896102674787035 --> Coincide con Numpy
Mediana = 4.650000095367432 --> Coincide con Numpy
Desviación estandar = 3.6895738135753655 --> No Coincide con Numpy, hay una diferencia de 0.0035122081380483117
Varianza = 13.612954925821066 --> No Coincide con Numpy, hay una diferencia de 0.02592943795394298
Coeficiente Variación = 0.6257648513063981 --> No Coincide con Numpy, hay una diferencia de 0.0005956830014286485
___________________________________________
 
Estadisticas educ
Media = 12.562737642585551 --> Coincide con Numpy
Mediana = 12.0 --> Coincide con Numpy
Desviación estandar = 2.766388999158773 --> No Coincide con Numpy, hay una diferencia de 0.00263340278492441
Varianza = 7.652908094666679 --> No Coincide con Numpy, hay una diferencia de 0.014576967799365015
Coeficiente Variación = 0.220205903988728 --> No Coincide con Numpy, hay una diferencia de 0.0002096201369355677
___________________________________________
 


### **Analizis resultados:** 


*   **Variable wage**: 

| Estadistico  |Valor     |
|--------------|----------|
| Media        |  5.89    |
| Mediana      |  4.65    |
| Desviación.E |  3.69    |
| Varianza     |  13.64   | 
| Coeficiente.V|  0.63    |
<Br>

Las ganancia promedio por hora (salario) de la muesta de población americana para el año de 1976 era de 5.89 USD, notese que esta cantidad de dinero o superior era percibida por menos de la mitad de los individuos puesto que la mediana de los datos es de apenas 4.65 USD. 

Asi mismo se puede comentar que en promedio las ganancias por hora de los individuos se desvian 3.693 USD de la media, lo cual muestra un alto nivel de variación en los salarios que se pueden confirmar interpretando el coeficiente de variación de los datos que haciende hasta un 0.62. 

Esto se puede succeder debido alguna fijacion de salario minimo que afecta a parte de los individuos mientras que el resto con mejores puestos logran obtener salarios mas altos. 
<Br>

* **Variable educ**:

| Estadistico  |Valor      |
|--------------|-----------|
| Media        |  12.56    |
| Mediana      |  12       |
| Desviación.E |  2.77     |
| Varianza     |  7.67     | 
| Coeficiente.V|  0.22     |
<Br>

Los años de educacuin promedio de la muestra de población americana para el año 1976 es de 12.65 años, notese que la mitad de los individuos por lo menos contaban con 12 años de educación.

Por otro lado se puede notar como en promedio los años de educacion de los individuos se desviaban de la media en 2.77 años, lo cual nos indica un bajo nivel de dispersion, esto puede ser causado por el hecho de que algunos individuos lograron terminar sus ciclos de primaria y bachillerato mientras que otros lograron adicionar algunos años de educacion extra en estudios profesionales.

Aun asi el coeficiente de variación de los años de educación es apenas de 0.22, lo que tambien es un indicativo de que el sistema educativo y gobierno del momento lograba que la diferencia en años de educación de los individuos no fuera tan alta.





